modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala

```diff
@@ -29,86 +30,16 @@ class Engine[D, U](stateL: Lens[D, Engine.State]) {
   type EventType = Event[ConcreteTypes]
   type ResultType = EventResult[ConcreteTypes]
   type UserEventType = UserEvent[ConcreteTypes]
+  type HandleType[A] = HandleP[D, EventType, A]
 
-  /**
-    * Type constructor where all Seqexec side effect are managed.
-    *
-    * It's named `Handle` after `fs2.Handle` in order to give a hint in a future
-    * migration.
-    */
-  type Handle[A] = HandleStateT[IO, A]
-  // Helper alias to facilitate lifting.
-  type HandleStateT[M[_], A] = StateT[M, D, A]
-
-  /*
-   * HandleP is a Stream which has as a side effect a State machine inside a IO, which can produce other
-   * Streams as output.
-   *
-   * Its type parameters are:
-   * A: Type of the output (usually Unit)
-   * D: Type of the user data included in the state machine state.
-   *
-   * Making it final causes an error: "The outer reference in this type test cannot be checked at run time"
-   */
-  case class HandleP[A](run: Handle[(A, Option[Stream[IO, EventType]])])
-  object HandleP {
-    def fromStream(p: Stream[IO, EventType]): HandleP[Unit] = {
-      HandleP[Unit](Applicative[Handle].pure[(Unit, Option[Stream[IO, EventType]])](((), Some(p))))
-    }
-  }
-
-  implicit def handlePInstances: Monad[HandleP] = new Monad[HandleP] {
-    private def concatOpP[F[_]](op1: Option[Stream[F, EventType]],
-                          op2: Option[Stream[F, EventType]]): Option[Stream[F, EventType]] = (op1, op2) match {
-      case (None, None)         => None
-      case (Some(p1), None)     => Some(p1)
-      case (None, Some(p2))     => Some(p2)
-      case (Some(p1), Some(p2)) => Some(p1 ++ p2)
-    }
-
-    override def pure[A](a: A): HandleP[A] = HandleP(Applicative[Handle].pure((a, None)))
-
-    override def flatMap[A, B](fa: HandleP[A])(f: A => HandleP[B]): HandleP[B] = HandleP[B](
-      fa.run.flatMap {
-        case (a, op1) => f(a).run.map{
-          case (b, op2) => (b, concatOpP(op1, op2))
-        }
-      }
-    )
-
-    // Kudos to @tpolecat
-    def tailRecM[A, B](a: A)(f: A => HandleP[Either[A, B]]): HandleP[B] = {
-      // We don't really care what this type is
-      type Unused = Option[Stream[IO, EventType]]
-
-      // Construct a StateT that delegates to IO's tailRecM
-      val st: StateT[IO, D, (B, Unused)] =
-        StateT { s =>
-          Monad[IO].tailRecM[(D, A), (D, (B, Unused))]((s, a)) {
-            case (s, a) =>
-              f(a).run.run(s).map {
-                case (sʹ, (Left(a), _))  => Left((sʹ, a))
-                case (sʹ, (Right(b), u)) => Right((sʹ, (b, u)))
-              }
-          }
-        }
-
-      // Done
-      HandleP(st)
-    }
-  }
-
-  implicit class HandleToHandleP[A](self: Handle[A]) {
-    def toHandleP: HandleP[A] = HandleP(self.map((_, None)))
-  }
```

Comment:
Moved `HandleP` to the companion object to make it accessible to other classes. Should it be in its own source file?

Reply:
probably. our source files tend to be too large